lib/SPIRV/SPIRVWriter.cpp

```diff
@@ -1118,6 +1121,38 @@ void LLVMToSPIRVBase::transFPGAFunctionMetadata(SPIRVFunction *BF,
     transMetadataDecorations(FDecoMD, BF);
 }
 
+void LLVMToSPIRVBase::transFunctionMetadataAsExecutionMode(SPIRVFunction *BF,
+                                                           Function *F) {
+  if (!BM->isAllowedToUseExtension(ExtensionID::SPV_INTEL_maximum_registers))
+    return;
+
+  if (auto *RegisterAllocModeMD = F->getMetadata("RegisterAllocMode")) {
+    auto *RegisterAllocMode = RegisterAllocModeMD->getOperand(0).get();
+    if (isa<MDString>(RegisterAllocMode)) {
+      StringRef Str = getMDOperandAsString(RegisterAllocModeMD, 0);
+      if (Str == "AutoINTEL")
+        BF->addExecutionMode(BM->add(new SPIRVExecutionMode(
+            OpExecutionMode, BF,
+            internal::ExecutionModeNamedMaximumRegistersINTEL, 0)));
+    } else if (isa<MDNode>(RegisterAllocMode)) {
+      auto *RegisterAllocNodeMDOp =
+          getMDOperandAsMDNode(RegisterAllocModeMD, 0);
+      int Num = getMDOperandAsInt(RegisterAllocNodeMDOp, 0);
+      auto *Const =
+          BM->addConstant(transType(Type::getInt32Ty(F->getContext())), Num);
+      BF->addExecutionMode(BM->add(new SPIRVExecutionMode(
+          OpExecutionModeId, BF, internal::ExecutionModeMaximumRegistersIdINTEL,
+          Const->getId())));
+    } else {
+      int64_t RegisterAllocNodeMDOp =
```

Comment:
Can we please rename this to differentiate with variable used in Line 1138?
May be RegisterAllocModeId?